### PR TITLE
Darwin: Use undocumented getgroupcount() before getgrouplist

### PR DESCRIPTION
**Pull request by russellhancox**
This fixes group lookups on OS X when a user is in more than 64 groups.

While the getgroupcount function is undocumented and not included in any headers, it has been exported from libSystem.B (via libsystem_info) since 10.6 and is used by the getgroups function, so it's very unlikely to disappear anytime soon.

### CODE REVIEW

**Comment by osqueryer** [comment]
Can one of the osquery admins reply with "ok to test" to kick off a build...

File: osquery/tables/system/user_groups.h [hunk 1]
```diff
@@ -50,6 +54,16 @@ struct user_t {
 template <typename uid_type, typename gid_type>
 static void getGroupsForUser(QueryData &results,
                              const user_t<uid_type, gid_type> &user) {
+#ifdef __APPLE__
+  int ngroups = getgroupcount(user.name, user.gid);
```
Comment (sharvilshah) [inline osquery/tables/system/user_groups.h]: Does it make sense to check if `ngroups <= 0` and bail early?

**Comment by sharvilshah** [comment]
Do you think we can get away with setting `_DARWIN_UNLIMITED_GETGROUPS` or `_DARWIN_C_SOURCE` like [`getgroups(2)`](https://developer.apple.com/library/mac/documentation/Darwin/Reference/ManPages/man2/getgroups.2.html) says? Its implementation is [here](https://opensource.apple.com/source/Libc/Libc-1082.20.4/sys/getgroups.c).

**Comment by russellhancox** [comment]
Those only apply to the getgroups function, they don't affect getgroupcount or getgrouplist

File: osquery/tables/system/user_groups.h (hunk 1 above)
Comment (russellhancox) [inline osquery/tables/system/user_groups.h]: I don't think that can happen in normal circumstances and the code works fine if it does happen, so I think that's unnecessary. I don't feel particularly strongly though.

**Comment by sharvilshah** [comment]
Right, but `getgroups()` calls `getgroupcount`/`getgrouplist`:

```c
/*
 * getgroups extension; not limited by NGROUPS_MAX
 */
int
getgroups(int gidsetsize, gid_t grouplist[])
{
    struct passwd *pw;
    int n;

    if ((pw = getpwuid(getuid())) == NULL) {
	  errno = EINVAL;
	  return -1;
    }
    if (gidsetsize == 0) {
	  if ((n = getgroupcount(pw->pw_name, pw->pw_gid)) == 0) {
	      errno = EINVAL;
	      return -1;
	  }
	  return n;
    }
    n = gidsetsize;
    if (getgrouplist(pw->pw_name, pw->pw_gid, (int *)grouplist, &n) < 0) {
	  errno = EINVAL;
	  return -1;
    }
    return n;
}
```

And from the manpage: `If gidsetsize is 0, getgroups() returns the number of groups without modifying
     the grouplist[] array.`

So I think, if `_DARWIN_UNLIMITED_GETGROUPS` or `_DARWIN_C_SOURCE` are honored, we might be able to get away with using `getgroups` instead of the undocumented API `getgroupcount`

**Comment by russellhancox** [comment]
Ah, I misunderstood.

getgroups can't be used because it always gets groups for the current user.

**Comment by sharvilshah** [comment]
Gotcha, and I don't think there is a public api that takes in either name/uid/pwd.

**Comment by theopolis** [comment]
ok to test

**Comment by osqueryer** [comment]
Build origin/pr/1884/merge (Job: 2420) is good! :+1:

**Comment by theopolis** [comment]
Just curious, are there anecdotes where a user is in 64+ groups? Maybe someone got a bit group happy in Active Directory?

**Comment by russellhancox** [comment]
I'm in >150, no Active Directory but something similar.

**Comment by theopolis** [comment]
:+1:

**Pull request merged**